### DISCUSSION TITLE
Group Hierarchy Confussion Clarification

### DISCUSSION THREAD

**Discussion by sridhar1495** [keycloak/keycloak#20341] [Q&A] [answered]
Configuration:
---------------
I am trying to configure fine grained access using groups in the KeyCloak,
As part of configuration, I have created following hierarchy with the user and roles,
`Manager (users: Rajesh, Roles: MANAGER) `
`    Leads (users: Jhon, Roles: TL) `
`        TeamA (Users: Amy, Roles: DEVOPS) `
`        TeamB (users: Kamal, Roles: DEV) `

Role Hierarchy:
----------------
The Hierarchy for the Role inheritance is I am expecting is that the lower-level group roles to be inherited to the higher-level roles.
i.e., as follows,
`Manager (Roles: [Assigned: MANAGER, Inherited: {TL, DEVOPS, DEV}]) `
`    Leads (Roles: [Assigned: TL, Inherited: {DEVOPS, DEV}]) `
`        TeamA (Roles: [Assigned: DEVOPS, Inherited: {}]) `
`        TeamB (Roles: [Assigned: DEV, Inherited: {}]) `

But the keycloak is inheriting the higher-level group roles to the lower-level groups.
i.e., as follows,
`Manager (Roles: [Assigned: MANAGER, Inherited: {}]) `
`    Leads (Roles: [Assigned: TL, Inherited: {MANAGER}]) `
`        TeamA (Roles: [Assigned: DEVOPS, Inherited: {MANAGER, TL}]) `
`        TeamB (Roles: [Assigned: DEV, Inherited: {MANAGER, TL}]) `

User Hierarchy:
-----------------
In the opposite direction I am expecting the users to be added to the lower-level groups from the higher-level groups so that they will be part of the child groups also. i.e., as follows,
`Manager (users: [Assigned: Rajesh, Inherited: {}]) `
`    Leads (users: [Assigned: Jhon, Inherited: {Rajesh}]) `
`        TeamA (users: [Assigned: Amy, Inherited: {Rajesh, Jhon}]) `
`        TeamB (users: [Assigned: Kamal, Inherited: {Rajesh, Jhon}]) `

But the keycloak is having the option of adding the subgroup users to the higher-level groups not what I am expecting.
i.e., as follows,
`Manager (users: [Assigned: Rajesh, Inherited: {Jhon, Amy, Kamal}]) `
`    Leads (users: [Assigned: Jhon, Inherited: {Amy, Kamal}]) `
`        TeamA (users: [Assigned: Amy, Inherited: {}]) `
`        TeamB (users: [Assigned: Kamal, Inherited: {}]) `

So, someone can give solution is there way to configure what I am expecting in the KeyCloak or do we need to raise this as a bug/enhancement?

**Comment by sschu** [comment, marked as answer]
The way this works is that roles are composites. If a user has a parent role, it automatically also has all its subroles which are the components of that parent role. For groups it is the other way round. If you are in a subgroup, you are automatically also in its parent group. This is pretty much how it is designed and it works as designed.
For me it looks like your group hierarchy is not correct because Leads are not a subgroup of Manager (it is rather the other way round).

**Reply by sridhar1495** [reply]
Thank you for the clarification. My fault that I got confused with another concept and misunderstood this. Now I am able to get what I need to do with group hierarchy.

**Comment by doanthai** [comment]
@sschu Do has any configuration for inverted pyramid structure or reverse child group?

**Reply by sschu** [reply]
I don't understand your question.

**Reply by doanthai** [reply]
@sschu I just want to configuration keycloak group role like the question. The more higher group has more permissions like  inverted pyramid structure and more lower group has less permissions.

**Reply by sschu** [reply]
Composite roles work that way, i.e. somebody who has a composite role can do more than somebody who only has one of the subroles.
